contracts/FungibleToken.cdc

```diff
@@ -102,136 +112,116 @@ pub contract interface FungibleToken {
-        /// @return A structure representing the requested view.
-        ///
-        pub fun resolveView(_ view: Type): AnyStruct? {
-            return nil
+        /// Checks if the given type is supported by this Vault
+        access(all) view fun isSupportedVaultType(type: Type): Bool {
+            return self.getSupportedVaultTypes()[type] ?? false
         }
-    }
-
-    /// The resource that contains the functions to send and receive tokens.
-    /// The declaration of a concrete type in a contract interface means that
-    /// every Fungible Token contract that implements the FungibleToken interface
-    /// must define a concrete `Vault` resource that conforms to the `Provider`, `Receiver`,
-    /// and `Balance` interfaces, and declares their required fields and functions
-    ///
-    pub resource Vault: Provider, Receiver, Balance {
-
-        /// The total balance of the vault
-        pub var balance: UFix64
 
-        // The conforming type must declare an initializer
-        // that allows providing the initial balance of the Vault
-        //
-        init(balance: UFix64)
-
-        /// Subtracts `amount` from the Vault's balance
+        /// withdraw subtracts `amount` from the Vault's balance
         /// and returns a new Vault with the subtracted balance
         ///
-        /// @param amount: The amount of tokens to be withdrawn from the vault
-        /// @return The Vault resource containing the withdrawn funds
-        ///
-        pub fun withdraw(amount: UFix64): @Vault {
+        access(Withdraw) fun withdraw(amount: UFix64): @{Vault} {
             pre {
                 self.balance >= amount:
                     "Amount withdrawn must be less than or equal than the balance of the Vault"
             }
             post {
+                result.getType() == self.getType(): "Must return the same vault type as self"
                 // use the special function `before` to get the value of the `balance` field
                 // at the beginning of the function execution
                 //
                 self.balance == before(self.balance) - amount:
-                    "New Vault balance must be the difference of the previous balance and the withdrawn Vault"
+                    "New Vault balance must be the difference of the previous balance and the withdrawn Vault balance"
             }
         }
 
-        /// Takes a Vault and deposits it into the implementing resource type
+        /// deposit takes a Vault and adds its balance to the balance of this Vault
         ///
-        /// @param from: The Vault resource containing the funds that will be deposited
-        ///
-        pub fun deposit(from: @Vault) {
+        access(all) fun deposit(from: @{FungibleToken.Vault}) {
             // Assert that the concrete type of the deposited vault is the same
             // as the vault that is accepting the deposit
             pre {
                 from.isInstance(self.getType()): 
```

Comment:
I think this should be:
```
self.isSupportedVaultType(type: from.getType()): "..."
```

which IMO seems more consistent with supporting multiple vault types in a single implementation

Reply:
I think you might be misunderstanding what it means to support multiple vault types. By that, we mean that contracts should be able to define more than one vault type in their contract, but vaults can't be allowed to accept deposits from vaults of different types. that would break a lot of assumptions and get things mixed up.